akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PersistentBehaviorStashSpec.scala

```diff
@@ -0,0 +1,443 @@
+      ackProbe.expectMessage(Ack("inc-4"))
+      ackProbe.expectMessage(Ack("inc-5"))
+      ackProbe.expectMessage(Ack("inc-6"))
+      ackProbe.expectMessage(Ack("inc-7"))
+
+      finalState.value should ===(7)
+    }
+
+    "handle many stashed" in {
+      val c = spawn(counter(nextPid()))
+      val ackProbe = TestProbe[Ack]
+      val stateProbe = TestProbe[State]
+      val notUsedProbe = TestProbe[NotUsed]
+
+      (1 to 100).foreach { n ⇒
+        c ! Increment(s"inc-1-$n", ackProbe.ref)
+      }
+
+      (1 to 3).foreach { n ⇒
+        c ! Deactivate(s"deact-2-$n", ackProbe.ref)
+      }
+
+      (1 to 100).foreach { n ⇒
+        if (n % 10 == 0)
+          c ! Unhandled(notUsedProbe.ref)
+        c ! Increment(s"inc-3-$n", ackProbe.ref)
+      }
+
+      c ! GetValue(stateProbe.ref)
+
+      (1 to 5).foreach { n ⇒
+        c ! UpdateValue(s"upd-4-$n", n * 1000, ackProbe.ref)
+      }
+
+      (1 to 3).foreach { n ⇒
+        c ! Activate(s"act-5-$n", ackProbe.ref)
+      }
+
+      (1 to 100).foreach { n ⇒
+        c ! Increment(s"inc-6-$n", ackProbe.ref)
+      }
+
+      c ! GetValue(stateProbe.ref)
+
+      (6 to 8).foreach { n ⇒
+        c ! UpdateValue(s"upd-7-$n", n * 1000, ackProbe.ref)
+      }
+
+      (1 to 3).foreach { n ⇒
+        c ! Deactivate(s"deact-8-$n", ackProbe.ref)
+      }
+
+      (1 to 100).foreach { n ⇒
+        c ! Increment(s"inc-9-$n", ackProbe.ref)
+      }
+
+      (1 to 3).foreach { n ⇒
+        c ! Activate(s"act-10-$n", ackProbe.ref)
+      }
+
+      c ! GetValue(stateProbe.ref)
+      val value1 = stateProbe.expectMessageType[State](5.seconds).value
+      val value2 = stateProbe.expectMessageType[State](5.seconds).value
+      val value3 = stateProbe.expectMessageType[State](5.seconds).value
+
+      // verify the order
+
+      (1 to 100).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"inc-1-$n"))
+      }
+
+      (1 to 3).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"deact-2-$n"))
+      }
+
+      (1 to 5).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"upd-4-$n"))
+      }
+
+      ackProbe.expectMessage(Ack("act-5-1"))
+
+      (1 to 100).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"inc-3-$n"))
+      }
+
+      (2 to 3).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"act-5-$n"))
+      }
+
+      (1 to 100).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"inc-6-$n"))
+      }
+
+      (6 to 8).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"upd-7-$n"))
+      }
+
+      (1 to 3).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"deact-8-$n"))
+      }
+
+      ackProbe.expectMessage(Ack("act-10-1"))
+
+      (1 to 100).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"inc-9-$n"))
+      }
+
+      (2 to 3).foreach { n ⇒
+        ackProbe.expectMessage(Ack(s"act-10-$n"))
+      }
+
+      value1 should ===(100)
+      value2 should ===(5200)
+      value3 should ===(8100)
+    }
+
+    // FIXME what is desired behavior?
```

Comment:
We need to decide if the stash buffers should be cleared or not for restarts.

Note that there are two causes of restarts:
* exception thrown when processing command, handled by outer supervision
* persist failures, handled by `onPersistFailure` backoff supervision

Reply:
This is tricky and I can't decide which way I prefer.

What are the use cases we see stashing being used?
* External resource being down?
* FSM and stashing events until we reach a later state

For `onPersistFailure` I'd likely want to stash to be preserved in both of those cases as we'd restore our state / stash to the same as before the database issue.

For a user exception it I lean the other way as that typically implies we've gotten into a bad state some how. Do I want to keep stashed events that got me there?

When are side effects run like replying to the user? If that is an Ack then it would be surprising if the eventual command were then lost, but the program could crash / drop in the mean time so there's still no guarantee. 

Another option for the user is to model a buffer directly into the state which I think covers the general FSM case.

Reply:
Maybe @renatocaval has some thoughts?

Reply:
Can that be overrideable with strategy?

Reply:
> For onPersistFailure I'd likely want to stash to be preserved in both of those cases as we'd restore our state / stash to the same as before the database issue.

I think I agree.

> For a user exception it I lean the other way as that typically implies we've gotten into a bad state some how.

I think I agree. For complete safe stashing the buffer of stashed commands must be persisted instead of using the in-memory buffer, i.e. implemented in application domain.

Reply:
@chbatey I don't know if I understand you in these two sentences:

> For a user exception it I lean the other way as that typically implies we've gotten into a bad state some how. Do I want to keep stashed events that got me there?

Why do you talk about stashed events? We stash the commands, no? As such, it's ok to preserve the stash after some user exception. The command that would have caused the exception won't be re-processed, but others that were waiting on the queue will have a chance to hit the command handlers. That is fine because we should not make any assumption about any relation between commands. 

If I have `CmdA` and `CmdB` in my mailbox/stash and `CmdA` causes an exception, then we let it crash and we deliver `CmdB` next. It may succeed or not, disregarding what happened with `CmdA`. 

> When are side effects run like replying to the user? If that is an Ack then it would be surprising if the eventual command were then lost, but the program could crash / drop in the mean time so there's still no guarantee.

That, and the previous sentence, makes me think that we want to keep a command around and re-deliver it in case of failures, until it succeeds. 

We could indeed keep it and retry if we can detect that that's only a DB error and not an user exception. That makes sense to me. However, we must be careful with how we deal with applying the events. 

As far as I remember, we do apply the events before we try to persist them. This is useful to check that we are not persisting anything that we can't apply later on. It's an extra guarantee. However, if the persist fails and we try to re-deliver the command we will emit and apply the events once again. 

If the model is immutable, we have protection against it. We can swap old and new state only when the persisting is confirmed, but the problem is that we can't know if the state is immutable.  We may have update it in place and the re-delivery of a failed command will screw up everything. 

Or do we really intent to let it crash and rehydrate the actor while preserving the stash? If that's the case, the state will be discarted and reconstructed on replay. And we will be safe.